File: lc-problems/437-Path-Sum-III/sample_dfs.py
#!/usr/bin/env python


class Solution:
    def pathSum(self, root: TreeNode, S: int) -> int:

        def dfs(node, sums, currSum):
            if not node:
                return
            newSum = node.val + currSum
            if newSum - S in sums:
                ans["num"] += sums[node.val + currSum - S]
            if not node.left and not node.right:
                return

            sums[newSum] = sums.get(newSum, 0) + 1
            dfs(node.left, sums, newSum)
            dfs(node.right, sums, newSum)
            sums[newSum] -= 1

        if not root:
            return 0

        ans = {"num": 0}
        sums = {0: 1}
        dfs(root, sums, 0)
        return ans["num"]

File: lc-problems/652-Find-Duplicate-Subtrees/sample_dfs.swift
class Solution {
    func findDuplicateSubtrees(_ root: TreeNode?) -> [TreeNode?] {
        var res:[TreeNode?] = []
        var map:[String:Int] = [:]
        helper(root, &res, &map)
        return res
    }
    
    func helper(_ root: TreeNode?, _ res: inout [TreeNode?], _ map: inout [String:Int]) -> String {
        guard let node = root else { return "#" }
        let text = node.val.description + "," + helper(node.left, &res, &map) + "," + helper(node.right, &res, &map)
        if map[text] == nil{
            map[text] = 1
        }else{
            if map[text] == 1{
                res.append(node)
                map[text]! += 1
            }
        }
        return text
    }
}